src/main/java/org/apache/sysml/hops/rewrite/HopDagValidator.java

```diff
@@ -28,85 +30,98 @@
 import org.apache.sysml.hops.Hop;
 import org.apache.sysml.hops.HopsException;
 import org.apache.sysml.hops.LiteralOp;
+import org.apache.sysml.parser.Expression;
 import org.apache.sysml.runtime.DMLRuntimeException;
 import org.apache.sysml.utils.Explain;
 
+import static org.apache.sysml.hops.HopsException.check;
+
 /**
  * This class allows to check hop dags for validity, e.g., parent-child linking.
- * It purpose is soley for debugging purposes (enabled in ProgramRewriter).
- * 
+ * Use it for debugging (enabled in {@link ProgramRewriter}).
  */
-public class HopDagValidator 
-{
+public class HopDagValidator {
 	private static final Log LOG = LogFactory.getLog(HopDagValidator.class.getName());
+
+	private HopDagValidator() {}
 	
-	public static void validateHopDag(ArrayList<Hop> roots) 
-		throws HopsException
-	{
+	public static void validateHopDag(final ArrayList<Hop> roots) throws HopsException {
 		if( roots == null )
 			return;
-		try
-		{
+		try {
 			Hop.resetVisitStatus(roots);
+			ValidatorState state = new ValidatorState();
 			for( Hop hop : roots )
-				rValidateHop(hop);
+				rValidateHop(hop, state);
 		}
-		catch(HopsException ex)
-		{
+		catch(HopsException ex) {
 			try {
 				LOG.error( "\n"+Explain.explainHops(roots) );
 			}catch(DMLRuntimeException e){}
-			
 			throw ex;
 		}
 	}
 	
-	public static void validateHopDag(Hop root) 
-		throws HopsException
-	{
+	public static void validateHopDag(final Hop root) throws HopsException {
 		if( root == null )
 			return;
-		
-		try
-		{
+		try {
 			root.resetVisitStatus();
-			rValidateHop(root);
+			ValidatorState state = new ValidatorState();
+			rValidateHop(root, state);
 		}
-		catch(HopsException ex)
-		{
+		catch(HopsException ex) {
 			try {
 				LOG.error( "\n"+Explain.explain(root) );
 			}catch(DMLRuntimeException e){}
-			
 			throw ex;
 		}
 	}
+
+	private static class ValidatorState {
+		final Set<Long> seen = new HashSet<>();
+	}
 	
-	private static void rValidateHop( Hop hop ) 
-		throws HopsException
-	{
-		if(hop.isVisited())
-			return;
+	private static void rValidateHop(final Hop hop, final ValidatorState state) throws HopsException {
+		final long id = hop.getHopID();
+
+		final boolean seen = !state.seen.add(id);
+		check(seen == hop.isVisited(), hop,
+				"seen previously is %b but hop visited previously is %b", seen, !seen);
```

Comment:
Please simplify this message - providing both inputs is not useful because 'seen' implies '!seen'.